File: VB/WindowsFormsApp4/Form1.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Diagram.Core
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Reflection
Imports System.Windows.Forms

Namespace WindowsFormsApp4

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
            CreateArcPoint.Register()
            RegisterStencil()
        End Sub

        Private Sub RegisterStencil()
            Dim stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("WindowsFormsApp4.CustomShapes.xml")
            Dim stencil = DiagramStencil.Create("CustomStencil", "Custom Shapes", stream, Function(shapeName) shapeName)
            DiagramToolboxRegistrator.RegisterStencil(stencil)
            diagramControl1.SelectedStencils = New StencilCollection() From {"CustomStencil"}
        End Sub
    End Class

    Public Class CreateArcPoint
        Implements ICustomFunctionOperator

        Private Shared ReadOnly instance As CreateArcPoint = New CreateArcPoint()

        Public Shared Sub Register()
            CriteriaOperator.RegisterCustomFunction(instance)
        End Sub

        Public Shared Sub Unregister()
            CriteriaOperator.UnregisterCustomFunction(instance)
        End Sub

        Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
            Get
                Return NameOf(CreateArcPoint)
            End Get
        End Property

        Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
            Return GetType(Double)
        End Function

        Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
            If operands.Length = 3 AndAlso TypeOf operands(0) Is Double AndAlso TypeOf operands(1) Is Double AndAlso TypeOf operands(2) Is String Then
                Dim X = CDbl(operands(0))
                Dim Y = CDbl(operands(1))
                Dim axe = CStr(operands(2))
                If axe Is "X" Then Return If(X > 0.5, X, 0.5)
                Return If(X > 0.5, Y, 1)
            Else
                Return Nothing
            End If
        End Function
    End Class
End Namespace

File: CS/WindowsFormsApp4/Form1.cs
using DevExpress.Data.Filtering;
using DevExpress.Diagram.Core;
using DevExpress.Utils.Serializing;
using DevExpress.XtraDiagram;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApp4
{
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
            CreateArcPoint.Register();
            RegisterStencil();
        }

        void RegisterStencil()
        {
            var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("WindowsFormsApp4.CustomShapes.xml");
            var stencil = DiagramStencil.Create("CustomStencil", "Custom Shapes", stream, shapeName => shapeName);
            DiagramToolboxRegistrator.RegisterStencil(stencil);

            diagramControl1.SelectedStencils = new StencilCollection() { "CustomStencil" };
        }
    }

    public class CreateArcPoint : ICustomFunctionOperator {
        private static readonly CreateArcPoint instance = new CreateArcPoint();

        public static void Register() {
            CriteriaOperator.RegisterCustomFunction(instance);
        }
        public static void Unregister() {
            CriteriaOperator.UnregisterCustomFunction(instance);
        }

        public string Name => nameof(CreateArcPoint);

        public Type ResultType(params Type[] operands) {
            return typeof(double);
        }

        public object Evaluate(params object[] operands) {
            if (operands.Length == 3
                && operands[0] is double
                && operands[1] is double
                && operands[2] is string) {

                var X = (double)operands[0];
                var Y = (double)operands[1];
                var axe = (string)operands[2];

                if (axe is "X")
                    return X > 0.5 ? X : 0.5;
                return X > 0.5 ? Y : 1;
            }
            else
                return null;
        }
    }
}
